src/muz/spacer/spacer_global_generalizer.cpp

```diff
@@ -168,48 +195,59 @@ static rational get_lcm(expr *e, ast_manager &m) {
     return g.m_val;
 }
 
-/// Converts all numerals and uninterpreted constants in fml to int
-/// requires that \p fml is in sop form
-static void to_int(expr_ref &fml) {
-    ast_manager &m = fml.get_manager();
-    arith_util arith(m);
-    TRACE("subsume_verb", tout << "to int " << mk_pp(fml, m) << "\n";);
-    if (arith.is_to_real(fml)) {
-        fml = to_app(fml)->get_arg(0);
-        TRACE("subsume_dbg_verb",
-              tout << "to int finished " << mk_pp(fml, m) << "\n";);
-        return;
-    }
-    if (arith.is_to_int(fml)) {
-        expr_ref res(m);
-        fml = to_app(fml)->get_arg(0);
-        to_int(fml);
-        return;
-    }
-    // Don't normalize constants.
-    if (is_uninterp_const(fml)) return;
-
-    rational val;
-    if (arith.is_numeral(fml, val)) {
-        // If its not an integer, try constructing int from it
-        fml = arith.mk_int(val);
-
-        TRACE("subsume_verb",
-              tout << "to int finished " << mk_pp(fml, m) << "\n";);
-        return;
-    }
-    app *fml_app = to_app(fml);
-    unsigned N = fml_app->get_num_args();
-    expr_ref_vector nw_args(m);
-    for (unsigned i = 0; i < N; i++) {
-        expr_ref chld(fml_app->get_arg(i), m);
-        to_int(chld);
-        nw_args.push_back(chld);
+struct strip_to_real_rw_cfg : public default_rewriter_cfg {
+    ast_manager &m;
+    arith_util m_arith;
+    array_util m_array;
+    strip_to_real_rw_cfg(ast_manager &m) : m(m), m_arith(m), m_array(m) {}
+    br_status reduce_app(func_decl *f, unsigned num, expr *const *args,
+                         expr_ref &result, proof_ref &result_pr) {
+        app *fml = nullptr;
+        if (num > 0)
+            // this call to mk_app creates a function whose domain has the same
+            // sort as args
+            fml = m.mk_app(f->get_family_id(), f->get_decl_kind(), num, args);
+        else
+            // create const using same f
+            fml = m.mk_const(f);
+        if (m_arith.is_to_real(fml)) {
+            result = fml;
+            return BR_DONE;
+        }
+        bool is_int;
+        rational r;
+        if (m_arith.is_numeral(fml, r, is_int) && !is_int) {
+            TRACE("global",
+                  tout << "found a rational when expecting an int\n";);
+            UNREACHABLE();
+            return BR_FAILED;
+        }
+        if (m_arith.is_to_int(fml) && m_arith.is_int(args[0])) {
+            result = args[0];
+            return BR_REWRITE_FULL;
```

Comment:
comment why FULL rewrite is needed. It is very expensive.

I expect that at this point `args[0]` is already rewritten, why is another descend over it is necessary? Why is there no bound on recursion? (i.e., rewrite is FULL?)